pkg/controller/direct/colab/runtimetemplate_controller.go

```diff
@@ -0,0 +1,284 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// +tool:controller
+// proto.service: google.cloud.aiplatform.v1beta1.NotebookService
+// proto.message: google.cloud.aiplatform.v1beta1.NotebookRuntimeTemplate
+// crd.type: ColabRuntimeTemplate
+// crd.version: v1alpha1
+
+package colab
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"strings"
+
+	gcp "cloud.google.com/go/aiplatform/apiv1beta1"
+	pb "cloud.google.com/go/aiplatform/apiv1beta1/aiplatformpb"
+	"google.golang.org/api/option"
+	"google.golang.org/protobuf/types/known/fieldmaskpb"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/klog/v2"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	krm "github.com/GoogleCloudPlatform/k8s-config-connector/apis/colab/v1alpha1"
+	refs "github.com/GoogleCloudPlatform/k8s-config-connector/apis/refs/v1beta1"
+	"github.com/GoogleCloudPlatform/k8s-config-connector/pkg/config"
+	"github.com/GoogleCloudPlatform/k8s-config-connector/pkg/controller/direct"
+	"github.com/GoogleCloudPlatform/k8s-config-connector/pkg/controller/direct/directbase"
+	"github.com/GoogleCloudPlatform/k8s-config-connector/pkg/controller/direct/registry"
+)
+
+func init() {
+	registry.RegisterModel(krm.ColabRuntimeTemplateGVK, NewRuntimeTemplateModel)
+}
+
+func NewRuntimeTemplateModel(ctx context.Context, config *config.ControllerConfig) (directbase.Model, error) {
+	return &runtimeTemplateModel{config: *config}, nil
+}
+
+var _ directbase.Model = &runtimeTemplateModel{}
+
+type runtimeTemplateModel struct {
+	config config.ControllerConfig
+}
+
+func (m *runtimeTemplateModel) client(ctx context.Context, projectID string) (*gcp.NotebookClient, error) {
+	var opts []option.ClientOption
+
+	config := m.config
+
+	// Workaround for an unusual behaviour (bug?):
+	//  the service requires that a quota project be set
+	if !config.UserProjectOverride || config.BillingProject == "" {
+		config.UserProjectOverride = true
+		config.BillingProject = projectID
+	}
+
+	opts, err := config.RESTClientOptions()
+	if err != nil {
+		return nil, err
+	}
+
+	gcpClient, err := gcp.NewNotebookRESTClient(ctx, opts...)
+	if err != nil {
+		return nil, fmt.Errorf("building colabruntimetemplate client: %w", err)
+	}
+
+	return gcpClient, err
+}
+
+func (m *runtimeTemplateModel) AdapterForObject(ctx context.Context, reader client.Reader, u *unstructured.Unstructured) (directbase.Adapter, error) {
+	obj := &krm.ColabRuntimeTemplate{}
+	if err := runtime.DefaultUnstructuredConverter.FromUnstructured(u.Object, &obj); err != nil {
+		return nil, fmt.Errorf("error converting to %T: %w", obj, err)
+	}
+
+	id, err := krm.NewNotebookRuntimeTemplateIdentity(ctx, reader, obj)
+	if err != nil {
+		return nil, err
+	}
+
+	mapCtx := &direct.MapContext{}
+	desired := ColabRuntimeTemplateSpec_ToProto(mapCtx, &obj.Spec)
+	if mapCtx.Err() != nil {
+		return nil, mapCtx.Err()
+	}
```

Comment:
Suggest put it back to create/update to distinguish these two functions.

Reply:
Done. This code logic was generated.